Tracking orphaned stream ids

havaker opened the pull request:
## Description

Implementation of orphaned stream id tracking that is resistant to `Connection::send_request` cancellation.

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [ ] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [ ] PR description sums up the changes and reasons why they should be introduced.
- [ ] I added appropriate `Fixes:` annotations to PR description.

Fixes: #305

psarna commented on scylla/src/transport/connection.rs:
```diff
@@ -649,14 +696,25 @@ impl Connection {
 
+        // Dropping `notifier` (before calling `notifier.disable()`) will send a notification to
+        // `Connection::router`. This notification is then used to mark a `stream_id` associated
+        // with this request as orphaned and free associated resources.
```
Isn't the following scenario possible?
1. The runtime yields from this code right after submitting the request above (`self.submit_channel`) succeeded, but before the line below runs (`let notifier = ...`).
2. This code's caller wrapped it in tokio::timeout, which fires
3. The request was already sent with an allocated stream id, but it's not going to be orphaned when no response arrives, because we yielded before creating the notifier

havaker replied:
Notifier is created in the `Connection::writer`, and then send here, to `Connection::send_request`. If a send fails, the notifier is dropped on the `Connection::writer` side, which also triggers mechanism for stream id orphaning.

So the third point is invalid, the allocated stream id will be marked as orphaned.

psarna commented on scylla/src/transport/connection.rs:
```diff
@@ -747,6 +805,7 @@ impl Connection {
         // are carefully written in such a way that they do not hold the lock
         // across .await points. Therefore, it should not be too expensive.
         let handler_map = StdMutex::new(ResponseHandlerMap::new());
+        let (orphan_sender, orphan_receiver) = mpsc::unbounded_channel();
```
I remember that you specifically want to use an unbounded_channel in order to be able to push to it synchronously - please add a short comment here mentioning this design decision

psarna commented on scylla/src/transport/connection.rs:
```diff
@@ -957,22 +962,36 @@ impl Connection {
+                    }
+                }
+                Some(gsi) = orphan_receiver.recv() => {
+                    let mut lock = handler_map.try_lock().unwrap(); // Same as above
```
nitpick: lock is a terrible name for this variable, because it's not a lock anymore, it's a map; please change it to something more descriptive, preferably ending with a `_map`

piodul replied:
`lock` is of type [`MutexGuard`](https://doc.rust-lang.org/std/sync/struct.MutexGuard.html), which is a kind of a lock. Maybe `guard`, `map_guard` or something similar will be more clear, then?

psarna replied:
@piodul technically it's a lock that gets dereferenced automatically, but calling this variable a "lock" would be equivalent to calling a C++ variable, which is a pointer to something, a "pointer".  `guard` sounds similarly non-descriptive to me, `map_guard` is better, `handler_map_guard` is better yet.

havaker commented:
I don't like the current design of receiving things from channels twice in `Connection::send_request`. The need for this arose by using stream id (inside the `OrphanageNotifier`) allocated by `Connection::writer`. If we get rid of this need, and introduce a way to identify a request *before* it gets send (as a `Task`) to `Connection::writer`, a notifier could use this identifier instead of stream id to notify about orphanage (and we could get rid of receiving stream id from `oneshot::channel` in `send_request` - make `send_request` receive from channel only once).

The request identifier (`request_id`) generator could be easily implemented using some monotonically increasing 64bit atomic in `Connection`. Each `send_request` call would allocate new `request_id`, create `OrphanageNotifier` with it, and send `Task` as usual, but with `request_id` as additional field. On the `Connection::router` side, after a stream id allocation, there would be a mapping from `request_id` to `stream_id`, enabling the router (more precisely `Connection::orpaner`) to react to orphanage notifications containing `request_id` only.

@psarna @piodul what do you think? Is it worth implementing?

psarna commented:
> I don't like the current design of receiving things from channels twice in `Connection::send_request`. The need for this arose by using stream id (inside the `OrphanageNotifier`) allocated by `Connection::writer`. If we get rid of this need, and introduce a way to identify a request _before_ it gets send (as a `Task`) to `Connection::writer`, a notifier could use this identifier instead of stream id to notify about orphanage (and we could get rid of receiving stream id from `oneshot::channel` in `send_request` - make `send_request` receive from channel only once).
> 
> The request identifier (`request_id`) generator could be easily implemented using some monotonically increasing 64bit atomic in `Connection`. Each `send_request` call would allocate new `request_id`, create `OrphanageNotifier` with it, and send `Task` as usual, but with `request_id` as additional field. On the `Connection::router` side, after a stream id allocation, there would be a mapping from `request_id` to `stream_id`, enabling the router (more precisely `Connection::orpaner`) to react to orphanage notifications containing `request_id` only.
> 
> @psarna @piodul what do you think? Is it worth implementing?

It kind of sounds like splitting this new `GenerationalStreamId` into two separate entities, right? A new 64bit counter for uniquely identifying a request, and later a legacy 16 bit value to be compliant with what CQL wants. I think it sounds better than the `GenerationalStreamId` idea.

havaker commented:
v2:

- added unique request identifiers
- removed `GenerationalStreamId`
- orphan notifications are now based on `request_id` instead of `stream_id`
- `UnableToAllocStreamId` error propagation
- applied suggestions from review

I'll post v1 vs v2 benchmark later.

havaker commented:
I did a benchmark between the v2 (`havaker:orphans`) and v1 (`havaker:generational-orphans`). It seems that v2 version is slightly faster.

## Setup

The benchmark was carried out on a single desktop machine with one Cassandra node (`docker run --rm -it --name cassandra -p 9042:9042 -p 19042:19042 cassandra`).
I've cloned [pkolaczk/latte](https://github.com/pkolaczk/latte), disabled lto, added debug symbols, changed scylla driver version to `havaker:*orphans` and run `cargo flamegraph -- run workloads/basic/write.rn -t 2`.

## Results

### v2

```
CONFIG ═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
            Date            Wed, 15 Dec 2021                                                          
            Time            19:11:40 +0100                                                            
         Cluster            Test Cluster                                                              
   Cass. version            4.0.1                                                                     
            Tags                                                                                      
        Workload            write.rn                                                                  
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
         Threads                    2                                                                 
     Connections                    1                                                                 
     Concurrency     [req]        128                                                                 
        Max rate    [op/s]                                                                            
          Warmup       [s]                                                                            
              └─      [op]          1                                                                 
        Run time       [s]       60.0                                                                 
              └─      [op]                                                                            
        Sampling       [s]        1.0                                                                 
              └─      [op]                                                                            

SUMMARY STATS ══════════════════════════════════════════════════════════════════════════════════════════════════════════════
    Elapsed time       [s]     60.003                                                                 
        CPU time       [s]     59.130                                                                 
 CPU utilisation       [%]       12.3                                                                 
          Cycles      [op]    7088418                                                                 
          Errors      [op]          0                                                                 
              └─       [%]        0.0                                                                 
        Requests     [req]    7088418                                                                 
              └─  [req/op]        1.0                                                                 
            Rows     [row]          0                                                                 
              └─ [row/req]        0.0                                                                 
         Samples                   60                                                                 
Mean sample size      [op]     118140                                                                 
              └─     [req]     118140                                                                 
     Concurrency     [req]         90 ± 2                                                             
              └─       [%]         70                                                                 
      Throughput    [op/s]     118136 ± 20624                                                         
              ├─   [req/s]     118136 ± 20624                                                         
              └─   [row/s]          0 ± 0                                                             
 Mean cycle time      [ms]      1.900 ± 0.457                                                         
 Mean resp. time      [ms]      1.897 ± 0.457                                                         


RESPONSE TIMES [ms] ════════════════════════════════════════════════════════════════════════════════════════════════════════
          Min                       1 ± 0                                                             
           25                       1 ± 0                                                             
           50                       2 ± 0                                                             
           75                       2 ± 1                                                             
           90                       2 ± 1                                                             
           95                       3 ± 1                                                             
           98                       4 ± 1                                                             
           99                       5 ± 2                                                             
           99.9                    66 ± 5                                                             
           99.99                   69 ± 9                                                             
          Max                      73 ± 10                                                            

RESPONSE TIME DISTRIBUTION ═════════════════════════════════════════════════════════════════════════════════════════════════
── Resp. time [ms] ──  ────────────────────────────────────────────── Count ────────────────────────────────────────────────
     0.0 ...      0.1          0   0.00%  
     0.1 ...      0.2          0   0.00%  
     0.2 ...      0.5         50   0.00%  
     0.5 ...      1.0     489128   6.90%  ▪▪▪▪▪▪▪
     1.0 ...      2.2    5554887  78.37%  ▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪
     2.2 ...      4.6     884158  12.47%  ▪▪▪▪▪▪▪▪▪▪▪▪▪
     4.6 ...     10.0     119454   1.69%  ▪
    10.0 ...     21.5      18349   0.26%  
    21.5 ...     46.4       3919   0.06%  
    46.4 ...    100.0      18419   0.26%  
   100.0 ...    215.4         30   0.00%  
   215.4 ...    464.2         24   0.00%  

```

25 records from `perf report` with a search for `scylla::transport::connection`

```
+   11.24%     0.00%  tokio-runtime-w  latte  [.] scylla::transport::connection::Connection::execute::{{closure}} (inlined)
+    9.05%     0.00%  tokio-runtime-w  latte  [.] scylla::transport::connection::Connection::send_request::{{closure}} (inlined)
+    6.81%     0.00%  tokio-runtime-w  latte  [.] scylla::transport::connection::Connection::router::{{closure}} (inlined)
+    6.79%     0.00%  tokio-runtime-w  latte  [.] scylla::transport::connection::Connection::router::{{closure}}::{{closure}} (inlined)
+    4.43%     0.00%  tokio-runtime-w  latte  [.] scylla::transport::connection::Connection::reader::{{closure}} (inlined)
+    2.27%     0.00%  tokio-runtime-w  latte  [.] scylla::transport::connection::Connection::writer::{{closure}} (inlined)
+    1.60%     0.29%  tokio-runtime-w  latte  [.] scylla::transport::connection::Connection::parse_response
+    1.22%     0.18%  tokio-runtime-w  latte  [.] scylla::transport::connection::ResponseHandlerMap::lookup
+    1.13%     0.10%  tokio-runtime-w  latte  [.] scylla::transport::connection::Connection::alloc_stream_id
+    0.89%     0.17%  tokio-runtime-w  latte  [.] scylla::transport::connection::ResponseHandlerMap::allocate
     0.70%     0.40%  tokio-runtime-w  latte  [.] scylla::transport::connection_pool::NodeConnectionPool::connection_for_token
+    0.64%     0.00%  tokio-runtime-w  latte  [.] scylla::transport::connection_pool::NodeConnectionPool::with_connections (inlined)
+    0.59%     0.00%  tokio-runtime-w  latte  [.] core::ptr::drop_in_place<core::option::Option<alloc::sync::Arc<tokio::sync::oneshot::Inner<core::result::Result<scylla::transport::connection::TaskResponse,scylla::transport::errors::QueryError>>>>> (inlined)
+    0.59%     0.00%  tokio-runtime-w  latte  [.] core::ptr::drop_in_place<alloc::sync::Arc<tokio::sync::oneshot::Inner<core::result::Result<scylla::transport::connection::TaskResponse,scylla::transport::errors::QueryError>>>> (inlined)
     0.39%     0.00%  tokio-runtime-w  latte  [.] core::ptr::drop_in_place<alloc::sync::Arc<scylla::transport::connection::Connection>> (inlined)
     0.27%     0.00%  tokio-runtime-w  latte  [.] core::ptr::drop_in_place<tokio::sync::oneshot::Inner<core::result::Result<scylla::transport::connection::TaskResponse,scylla::transport::errors::QueryError>>> (inlined)
     0.25%     0.00%  tokio-runtime-w  latte  [.] scylla::transport::connection::OrphanageTracker::contains (inlined)
     0.22%     0.00%  tokio-runtime-w  latte  [.] core::ptr::drop_in_place<core::option::Option<core::result::Result<scylla::transport::connection::TaskResponse,scylla::transport::errors::QueryError>>> (inlined)
     0.18%     0.18%  tokio-runtime-w  latte  [.] scylla::transport::connection_pool::NodeConnectionPool::choose_random_connection_from_slice
     0.17%     0.14%  tokio-runtime-w  latte  [.] scylla::transport::connection::QueryResponse::into_query_result
     0.17%     0.00%  tokio-runtime-w  latte  [.] scylla::transport::connection_pool::NodeConnectionPool::connection_for_token::{{closure}} (inlined)
     0.14%     0.00%  tokio-runtime-w  latte  [.] core::ptr::drop_in_place<std::sync::mutex::MutexGuard<scylla::transport::connection::ResponseHandlerMap>> (inlined)
     0.12%     0.11%  tokio-runtime-w  latte  [.] scylla::transport::connection::Connection::allocate_request_id
     0.11%     0.00%  tokio-runtime-w  latte  [.] scylla::transport::connection::Connection::orphaner::{{closure}}::{{closure}} (inlined)
     0.10%     0.00%  tokio-runtime-w  latte  [.] core::ptr::drop_in_place<std::sync::mutex::MutexGuard<scylla::transport::connection::ResponseHandlerMap>> (inlined)
```

### v1

```
CONFIG ═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
            Date            Wed, 15 Dec 2021                                                          
            Time            19:19:20 +0100                                                            
         Cluster            Test Cluster                                                              
   Cass. version            4.0.1                                                                     
            Tags                                                                                      
        Workload            write.rn                                                                  
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
         Threads                    2                                                                 
     Connections                    1                                                                 
     Concurrency     [req]        128                                                                 
        Max rate    [op/s]                                                                            
          Warmup       [s]                                                                            
              └─      [op]          1                                                                 
        Run time       [s]       60.0                                                                 
              └─      [op]                                                                            
        Sampling       [s]        1.0                                                                 
              └─      [op]                                                                            

SUMMARY STATS ══════════════════════════════════════════════════════════════════════════════════════════════════════════════
    Elapsed time       [s]     60.003                                                                 
        CPU time       [s]     61.320                                                                 
 CPU utilisation       [%]       12.8                                                                 
          Cycles      [op]    6827263                                                                 
          Errors      [op]          0                                                                 
              └─       [%]        0.0                                                                 
        Requests     [req]    6827263                                                                 
              └─  [req/op]        1.0                                                                 
            Rows     [row]          0                                                                 
              └─ [row/req]        0.0                                                                 
         Samples                   60                                                                 
Mean sample size      [op]     113788                                                                 
              └─     [req]     113788                                                                 
     Concurrency     [req]         91 ± 2                                                             
              └─       [%]         71                                                                 
      Throughput    [op/s]     113785 ± 32071                                                         
              ├─   [req/s]     113785 ± 32071                                                         
              └─   [row/s]          0 ± 0                                                             
 Mean cycle time      [ms]      1.981 ± 0.690                                                         
 Mean resp. time      [ms]      1.977 ± 0.690                                                         


RESPONSE TIMES [ms] ════════════════════════════════════════════════════════════════════════════════════════════════════════
          Min                       1 ± 0                                                             
           25                       1 ± 0                                                             
           50                       2 ± 1                                                             
           75                       2 ± 1                                                             
           90                       3 ± 1                                                             
           95                       3 ± 1                                                             
           98                       4 ± 2                                                             
           99                       5 ± 3                                                             
           99.9                    67 ± 6                                                             
           99.99                   68 ± 6                                                             
          Max                      70 ± 6                                                             

RESPONSE TIME DISTRIBUTION ═════════════════════════════════════════════════════════════════════════════════════════════════
── Resp. time [ms] ──  ────────────────────────────────────────────── Count ────────────────────────────────────────────────
     0.0 ...      0.1          0   0.00%  
     0.1 ...      0.2          0   0.00%  
     0.2 ...      0.5         13   0.00%  
     0.5 ...      1.0     421154   6.17%  ▪▪▪▪▪▪
     1.0 ...      2.2    5249288  76.89%  ▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪▪
     2.2 ...      4.6     935312  13.70%  ▪▪▪▪▪▪▪▪▪▪▪▪▪▪
     4.6 ...     10.0     178430   2.61%  ▪▪
    10.0 ...     21.5      22082   0.32%  
    21.5 ...     46.4       2783   0.04%  
    46.4 ...    100.0      17880   0.26%  
   100.0 ...    215.4        321   0.00%  
```

25 records from `perf report` with a search for `scylla::transport::connection`

```
+   12.75%     0.00%  tokio-runtime-w  latte  [.] scylla::transport::connection::Connection::execute::{{closure}} (inlined)
+   10.69%     0.00%  tokio-runtime-w  latte  [.] scylla::transport::connection::Connection::send_request::{{closure}} (inlined)
+    6.97%     0.00%  tokio-runtime-w  latte  [.] scylla::transport::connection::Connection::router::{{closure}} (inlined)
+    6.95%     0.00%  tokio-runtime-w  latte  [.] scylla::transport::connection::Connection::router::{{closure}}::{{closure}} (inlined)
+    3.70%     0.00%  tokio-runtime-w  latte  [.] scylla::transport::connection::Connection::reader::{{closure}} (inlined)
+    3.17%     0.00%  tokio-runtime-w  latte  [.] scylla::transport::connection::Connection::writer::{{closure}} (inlined)
+    1.60%     0.26%  tokio-runtime-w  latte  [.] scylla::transport::connection::Connection::parse_response
+    1.02%     0.12%  tokio-runtime-w  latte  [.] scylla::transport::connection::Connection::alloc_generational_stream_id
     0.81%     0.38%  tokio-runtime-w  latte  [.] scylla::transport::connection::ResponseHandlerMap::allocate
+    0.66%     0.13%  tokio-runtime-w  latte  [.] scylla::transport::connection::ResponseHandlerMap::lookup
     0.62%     0.35%  tokio-runtime-w  latte  [.] scylla::transport::connection_pool::NodeConnectionPool::connection_for_token
+    0.57%     0.00%  tokio-runtime-w  latte  [.] core::ptr::drop_in_place<core::option::Option<alloc::sync::Arc<tokio::sync::oneshot::Inner<core::result::Result<scylla::transport::connection::TaskResponse,scylla::transport::errors::QueryError>>>>> (inlined)
+    0.57%     0.00%  tokio-runtime-w  latte  [.] core::ptr::drop_in_place<alloc::sync::Arc<tokio::sync::oneshot::Inner<core::result::Result<scylla::transport::connection::TaskResponse,scylla::transport::errors::QueryError>>>> (inlined)
+    0.55%     0.00%  tokio-runtime-w  latte  [.] scylla::transport::connection_pool::NodeConnectionPool::with_connections (inlined)
     0.48%     0.00%  tokio-runtime-w  latte  [.] core::ptr::drop_in_place<core::option::Option<alloc::sync::Arc<tokio::sync::oneshot::Inner<scylla::transport::connection::OrphanhoodNotifier>>>> (inlined)
     0.48%     0.00%  tokio-runtime-w  latte  [.] core::ptr::drop_in_place<alloc::sync::Arc<tokio::sync::oneshot::Inner<scylla::transport::connection::OrphanhoodNotifier>>> (inlined)
     0.39%     0.00%  tokio-runtime-w  latte  [.] scylla::transport::connection::ResponseHandlerMap::new_generation (inlined)
     0.38%     0.00%  tokio-runtime-w  latte  [.] core::ptr::drop_in_place<alloc::sync::Arc<scylla::transport::connection::Connection>> (inlined)
     0.26%     0.25%  tokio-runtime-w  latte  [.] scylla::transport::connection::OrphanhoodNotifier::disable
     0.24%     0.00%  tokio-runtime-w  latte  [.] core::ptr::drop_in_place<tokio::sync::oneshot::Inner<core::result::Result<scylla::transport::connection::TaskResponse,scylla::transport::errors::QueryError>>> (inlined)
     0.23%     0.00%  tokio-runtime-w  latte  [.] core::ptr::drop_in_place<scylla::transport::connection::OrphanhoodNotifier> (inlined)
     0.23%     0.00%  tokio-runtime-w  latte  [.] core::ptr::drop_in_place<tokio::sync::mpsc::unbounded::UnboundedSender<scylla::transport::connection::GenerationalStreamId>> (inlined)
     0.23%     0.00%  tokio-runtime-w  latte  [.] core::ptr::drop_in_place<tokio::sync::mpsc::chan::Tx<scylla::transport::connection::GenerationalStreamId,tokio::loom::std::atomic_usize::AtomicUsize>> (inlined)
     0.23%     0.00%  tokio-runtime-w  latte  [.] scylla::transport::connection::OrphanageTracker::contains (inlined)
```

### Flamegraphs

[flamegraphs.zip](https://github.com/scylladb/scylla-rust-driver/files/7721984/flamegraphs.zip)

psarna reviewed (commented):
Sounds good

havaker commented on scylla/src/transport/connection.rs:
```diff
@@ -808,24 +911,32 @@ impl Connection {
+                    // Unsolicited frame. This should not happen and indicates
+                    // a bug either in the driver, or in the database
+                    return Err(QueryError::ProtocolError(
+                        "Received reponse with unexpected StreamId",
```
`ProtocolError` can only carry `&'static str` messages, so I'll log it.

havaker commented:
v3: added unexpected `StreamId` logging

havaker commented:
I run a test which spammed a Scylla instance on another machine with quickly timeouting queries (`session.query()` wrapped in `tokio::time::timeout` with a timeout duration lower than RTT). After appropriate reduction of `OLD_AGE_ORPHAN_THRESHOLD` i was able to observe a log message `Too many old orphaned stream ids`. Affected connection was closed and then reestablished by `connection_pool`.

Because the test was successful, I'm dropping `Draft` marker from this PR.

psarna reviewed (approved):
Looks good. Please also rerun the test which triggers orphaning the ids and generate a flamegraph out of the run with `cargo flamegraph`, and then post the flamegraph here. I'm interested in seeing if anything new pops out on the hot path (it shouldn't, but it never hurts to double-check).

havaker commented:
I've created a test, which does a lot requests concurrently and orphans ~half of them.
It can be seen here: https://gist.github.com/havaker/097521a9bb933f668b1630e683f56353.

I've run it on a 2-machine setup:
- powerful desktop which hosted a scylla instance 
- laptop on which the `bench-orphans.rs ` was run using `cargo flamegraph`

### Results:

Output:
```
Timed out queries count = 4180900 (49% of all queries)
```

25 records from `perf report` with a search for `scylla::transport::connection`:
```
+   29.27%     0.00%  tokio-runtime-w  orphan-bench  [.] scylla::transport::connection::Connection::router::{{closure}} (inlined)
+   28.91%     0.00%  tokio-runtime-w  orphan-bench  [.] scylla::transport::connection::Connection::router::{{closure}}::{{closure}} (inlined)
+   16.70%     0.00%  tokio-runtime-w  orphan-bench  [.] scylla::transport::connection::Connection::query::{{closure}} (inlined)
+   15.64%     0.00%  tokio-runtime-w  orphan-bench  [.] scylla::transport::connection::Connection::send_request::{{closure}} (inlined)
+   10.35%     0.00%  tokio-runtime-w  orphan-bench  [.] scylla::transport::connection::Connection::reader::{{closure}} (inlined)
+    9.12%     0.00%  tokio-runtime-w  orphan-bench  [.] scylla::transport::connection::Connection::writer::{{closure}} (inlined)
+    8.62%     0.00%  tokio-runtime-w  orphan-bench  [.] scylla::transport::connection::Connection::orphaner::{{closure}} (inlined)
+    4.10%     0.67%  tokio-runtime-w  orphan-bench  [.] scylla::transport::connection::ResponseHandlerMap::orphan
+    3.81%     0.00%  tokio-runtime-w  orphan-bench  [.] scylla::transport::connection::Connection::orphaner::{{closure}}::{{closure}} (inlined)
+    3.32%     1.03%  tokio-runtime-w  orphan-bench  [.] scylla::transport::connection::ResponseHandlerMap::lookup
+    3.05%     0.23%  tokio-runtime-w  orphan-bench  [.] core::ptr::drop_in_place<core::future::from_generator::GenFuture<scylla::transport::session::Session::run_query<core::future::from_generator::GenFuture<scylla::transport::session::Session::query_paged<&str,(i32,)>::{{closure}}::{{closure}}::{{closure}}>,core::future::from_generator::GenFuture<scylla::transport::session::Session::query_paged<&
+    3.00%     0.00%  tokio-runtime-w  orphan-bench  [.] core::ptr::drop_in_place<scylla::transport::session::Session::run_query<core::future::from_generator::GenFuture<scylla::transport::session::Session::query_paged<&str,(i32,)>::{{closure}}::{{closure}}::{{closure}}>,core::future::from_generator::GenFuture<scylla::transport::session::Session::query_paged<&str,(i32,)>::{{closure}}::{{closure}}::{
+    2.98%     0.22%  tokio-runtime-w  orphan-bench  [.] scylla::transport::connection::Connection::alloc_stream_id
+    2.96%     0.16%  tokio-runtime-w  orphan-bench  [.] scylla::transport::connection::Connection::parse_response
+    2.33%     0.49%  tokio-runtime-w  orphan-bench  [.] core::ptr::drop_in_place<core::future::from_generator::GenFuture<scylla::transport::session::Session::execute_query<core::future::from_generator::GenFuture<scylla::transport::session::Session::query_paged<&str,(i32,)>::{{closure}}::{{closure}}::{{closure}}>,core::future::from_generator::GenFuture<scylla::transport::session::Session::query_pag
+    2.27%     0.00%  tokio-runtime-w  orphan-bench  [.] core::ptr::drop_in_place<scylla::transport::session::Session::execute_query<core::future::from_generator::GenFuture<scylla::transport::session::Session::query_paged<&str,(i32,)>::{{closure}}::{{closure}}::{{closure}}>,core::future::from_generator::GenFuture<scylla::transport::session::Session::query_paged<&str,(i32,)>::{{closure}}::{{closure}
+    1.96%     0.15%  tokio-runtime-w  orphan-bench  [.] scylla::transport::connection::ResponseHandlerMap::allocate
+    1.81%     0.65%  tokio-runtime-w  orphan-bench  [.] scylla::transport::connection_pool::NodeConnectionPool::random_connection
+    1.80%     0.00%  tokio-runtime-w  orphan-bench  [.] scylla::transport::connection::OrphanageTracker::insert (inlined)
+    1.73%     0.00%  tokio-runtime-w  orphan-bench  [.] scylla::transport::connection_pool::NodeConnectionPool::with_connections (inlined)
+    1.70%     0.24%  tokio-runtime-w  orphan-bench  [.] core::ptr::drop_in_place<core::future::from_generator::GenFuture<scylla::transport::connection::Connection::query<&alloc::borrow::Cow<scylla::frame::value::SerializedValues>>::{{closure}}>>
+    1.63%     0.00%  tokio-runtime-w  orphan-bench  [.] core::ptr::drop_in_place<scylla::transport::connection::Connection::query<&alloc::borrow::Cow<scylla::frame::value::SerializedValues>>::{{closure}}> (inlined)
+    1.55%     0.21%  tokio-runtime-w  orphan-bench  [.] core::ptr::drop_in_place<core::future::from_generator::GenFuture<scylla::transport::connection::Connection::send_request<scylla::frame::request::query::Query>::{{closure}}>>
+    1.49%     0.00%  tokio-runtime-w  orphan-bench  [.] core::ptr::drop_in_place<scylla::transport::connection::Connection::send_request<scylla::frame::request::query::Query>::{{closure}}> (inlined)
+    1.17%     0.00%  tokio-runtime-w  orphan-bench  [.] core::ptr::drop_in_place<scylla::transport::connection::OrphanhoodNotifier> (inlined)

```
Flamegraph:

[flamegraph.zip](https://github.com/scylladb/scylla-rust-driver/files/7865817/flamegraph.zip)

psarna commented:
Also, if I parsed the flamegraph correctly, here's the overhead for managing orphans:

![orph](https://user-images.githubusercontent.com/10433047/149475262-e3b67b35-1a79-4a07-8526-8149a5f388a8.png)

A little sad to see the overhead, but on the other hand it's inevitable to have an impact with 49% of all queries timing out, and it's good to see that it didn't overwhelm the CPU at all. Sounds acceptable to me.

psarna commented:
side note: please create an issue about the fixmes - making the constants related to orphaning configurable
